File: src/algorithm/leetcode/Problem15.java
package algorithm.leetcode;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class Problem15 {
    public static List<List<Integer>> threeSum(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();
        Arrays.sort(nums);
        for (int i = 0; i < nums.length - 2; i++) {
            int first = i;
            int second = i + 1;
            int last = nums.length - 1;
            while (second < last) {
                int sum = nums[second] + nums[last];
                if (sum < -1 * nums[first]) {
                    second++;
                } else if (sum > -1 * nums[first]) {
                    last--;
                } else {
                    List<Integer> lst = new ArrayList<>();
                    lst.add(nums[first]);
                    lst.add(nums[second]);
                    lst.add(nums[last]);
                    result.add(lst);
                    while (second < nums.length - 1 && nums[second] == nums[second + 1]) {
                        second++;
                    }
                    second++;
                    while (last > 0 && nums[last] == nums[last - 1]) {
                        last--;
                    }
                    last--;
                }
            }
            while (i < nums.length - 1 && nums[i] == nums[i + 1]) {
                i++;
            }
        }
        return result;
    }

    public static void main(String[] args) {
        int[] tmp = {-1,0,1,2,-1,-4};
        List<List<Integer>> rst = threeSum(tmp);
        System.out.println();
    }
}

File: src/algorithm_cpp/problem15.cpp
#include<algorithm>
#include<iostream>
#include<vector>

using namespace std;

class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
	sort(nums.begin(), nums.end());
	vector<vector<int>> ans;
	for (int i = 0; i < (int)nums.size() - 2; i++) {
	    int l = i + 1;
	    int r = nums.size() - 1;
	    while(l < r) {
		if (nums[i] + nums[l] + nums[r] < 0) {
		    while(l < r && nums[l + 1] == nums[l]) 
			l++;
		    l++;
		} else if (nums[i] + nums[l] + nums[r] > 0) {
		    while(l < r && nums[r - 1] == nums[r])
			r--;
		    r--;
		} else {
		    vector<int> res {nums[i], nums[l], nums[r]};
		    ans.push_back(res);
		    while(l < r && nums[l + 1] == nums[l])
			l++;
		    l++;
		    while(l < r && nums[r - 1] == nums[r])
			r--;
		    r--;
		}
	    }
	    while(i < nums.size() - 2 && nums[i + 1] == nums[i])
		i++;
	}
	return ans;		
    }
};

int main() {
    //vector<int> nums {-1, 0, 1, 2, -1, -4};
    vector<int> nums;
    vector<vector<int>> ans;
    Solution s;
    cout << "call three sum" << endl;
    ans = s.threeSum(nums);
    cout << "end three sum" << endl;
    for (int i = 0; i < ans.size(); i++) {
	for (int j = 0; j < ans.at(i).size(); j++) {
	    cout << ans.at(i).at(j) << ", ";
	}
	cout << endl;
    }
}
